proposals/p1885.md

```diff
@@ -0,0 +1,710 @@
+```
+
+#### Interoperability with C++ types and containers
+
+The cursor approach deviates from the C++ iterator approach, requiring some
+adaptation.
+
+Given the current state of the design, C++ interoperability is yet to be
+defined, and suggestions have been highlighted in the
+[Future work](#future-work) section.
+
+## Future work
+
+### `for` with mutable values
+
+Supporting mutable values could be achieved using a second interface that
+provides a proxy or view for the data, exposing an `Iterate` interface with
+`ElementType*`
+
+```carbon
+// Object implementing `MutableIterate` returned by a function
+for (p: T* in my_range.AddrRange()) {}
+```
+
+The example `MutableIterate` interface below acts as a proxy for the collection,
+and implements our `Iterate` interface.
+
+```carbon
+interface MutableIterate {
+  impl as Iterate;
+  alias ElementType = Iterate.ElementType;
+  let ProxyType:! Iterate where .ElementType = ElementType*
+      and .CursorType is ImplicitAs((Self as Iterate).CursorType);
+  fn AddrRange[addr self: Self*]() -> ProxyType;
+}
+
+class MyIntContainer {
+  //...
+  impl as Iterate ...
+  external impl as MutateIterate where .ProxyType = Self {
+    ...
+  }
+}
+```
+
+### Speculative mixin usage
+
+Mixins are currently in early design stages. This section highlights possible
+uses speculating on the final design. Some may include:
+
+-   Improved semantics for views compared to getter methods: no direct side
+    effect from using the view
+-   Facilitate code reuse, compared to reimplementing an interface
+-   Direct access to `self`, limiting needs for pointers and address resolution
+
+#### Mutable values
+
+Named mixins, used as programmable members, could expose a view with mutable
+values:
+
+```carbon
+// Using a named mixin to expose mutable values
+for (p: my_range.(Iterate.ElementType)* in my_range.mutable) {}
+```
+
+Used in this way, this clarifies that this a view of the data, like an
+attribute, with no direct side effects, when compared to a method.
+
+#### Alternative views
+
+Similarly, mixins could allow supporting other views into the container.
+
+For example, maps could expose a view with "key" and "value" only, in addition
+to the default (key, value) tuple.
+
+```carbon
+// Using a named mixin to expose mutable values
+for (k: my_dict.(Iterate.CursorType) in my_dict.keys) {}
+```
+
+And "reversed" view could also be used for reverse iteration:
+
+```carbon
+// Reversing
+for (v: my_dict.(Iterate.ElementType) in my_dict.reversed) {}
+```
+
+### Large elements and Optional
+
+Large elements would have a negative impact if the `Optional` cannot leverage
+unused bit patterns within the type, nor present a view instead of a copy.
+
+Such opportunities for `Optional` exist, and could be addressed in the future.
+
+### C++ interoperability
+
+The limitation of this approach is the need to adapt to C++ iterator model. This
+could be implemented by having the cursor be an iterator type:
```

Comment:
There are two directions for interop. I think what you are saying is that we can create a Carbon cursor from a C++ iterator, but I think that should be clarified. It raises the question about interop in the other direction, which should ideally be addressed as well. Maybe there should be two subsections to make it very explicit?